editor/src/doc.rs

```diff
@@ -153,13 +171,66 @@ impl<'l> Doc<'l> {
         Ok(())
     }
 
-    fn execute_ed(&mut self, _cmd: EditorCmd) -> Result<UndoGroup<'l>, ()> {
-        unimplemented!();
+    fn execute_ed(
+        &mut self,
+        cmd: EditorCmd,
+        clipboard: &mut Vec<Ast<'l>>,
+    ) -> Result<UndoGroup<'l>, ()> {
+        if !self.mode.is_tree() {
+            panic!("tried to execute editor command in text mode")
+        }
+        match cmd {
+            EditorCmd::Cut => {
+                let (undos, ast) = self.remove(true)?;
+                clipboard.push(ast.expect("failed to cut"));
+                Ok(UndoGroup::with_edit(undos))
+            }
+            EditorCmd::Copy => {
+                clipboard.push(self.ast.clone());
+                Ok(UndoGroup::new())
+            }
+            EditorCmd::PasteAfter => {
+                if let Some(tree) = clipboard.pop() {
+                    // TODO if the insert fails, we'll lose the tree forever...
+                    self.execute_tree(TreeCmd::InsertAfter(tree))
+                } else {
+                    // TODO should we return an error if the clipboard is empty?
```

Comment:
Yes: error is for when you try to execute a command, but it can't be executed. The paste failed to execute.

Reply:
This is fixed in a later unpushed commit: it will return `DocError::EmptyClipboard` in this case.